### PR TITLE
[Store] feat: add tenant quota core

### PR DESCRIPTION
**Pull request by Lin-z-w**
## Description

This PR adds the standalone Dynamic Tenant Quota core for Mooncake Store.

It introduces `TenantQuotaTable` with policy management, explicit/default tenant quota recomputation, quota reservation/commit/abort/release accounting, tenant snapshots, and focused unit coverage. The recompute logic prioritizes tenants with explicit quota policies and lets default/unlimited tenants share only the remaining allocatable capacity, using largest-remainder distribution with stable tenant-id tie breaking.

This is intentionally not wired into `MasterService` lifecycle, HTTP/admin APIs, metrics, snapshots, or store-level `ErrorCode` mapping yet.

## Module

- [ ] Transfer Engine (`mooncake-transfer-engine`)
- [x] Mooncake Store (`mooncake-store`)
- [ ] Mooncake EP (`mooncake-ep`)
- [ ] Integration (`mooncake-integration`)
- [ ] P2P Store (`mooncake-p2p-store`)
- [ ] Python Wheel (`mooncake-wheel`)
- [ ] PyTorch Backend (`mooncake-pg`)
- [ ] Mooncake RL (`mooncake-rl`)
- [ ] CI/CD
- [ ] Docs
- [ ] Other

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other

## How Has This Been Tested?

- `cmake --build build -j$(nproc) --target tenant_quota_test`
- `ctest --test-dir build -R tenant_quota_test --output-on-failure`

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have formatted my own code using `./scripts/code_format.sh` before submitting.
- [ ] I have updated the documentation.
- [x] I have added tests to prove my changes are effective.

### CODE REVIEW

File: mooncake-store/src/tenant_quota.cpp [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+        return;
+    }
+    auto& state = it->second;
+    state.requested_quota_bytes = default_requested_quota_bytes_;
+    state.has_explicit_policy = false;
+}
+
+void TenantQuotaTable::RecomputeEffectiveQuotas(
+    uint64_t allocatable_capacity_bytes) {
+    unsigned __int128 explicit_requested_sum = 0;
```
[the remark this answers is not in the document]
Comment (Lin-z-w) [reply mooncake-store/src/tenant_quota.cpp:80]: Thanks, this is a valid correctness issue.

I fixed this by excluding lazy-empty inherited tenants from the default-capacity distribution in `RecomputeEffectiveQuotas()`. They still remain in the table as lazy state, but their `effective_quota_bytes` is reset to `0` and they no longer dilute active default tenants.

I also added a regression test covering an active default tenant plus a lazy-empty `ghost` tenant after policy erase.

File: mooncake-store/include/tenant_quota.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include <cstdint>
+#include <map>
```
[the remark this answers is not in the document]
Comment (Lin-z-w) [reply mooncake-store/include/tenant_quota.h:4]: Thanks for the suggestion. I’m going to keep `std::map` here for PR1.

This `TenantQuotaTable` is currently a standalone quota-core/test carrier and is not wired into the production transfer critical path. In the follow-up integration PR, the runtime quota state will be embedded into `MasterService::TenantState`, which already lives under the master service tenant map.

Keeping `std::map` also gives deterministic iteration order for snapshots and recompute tie-breaking without adding extra sorting code.

File: mooncake-store/include/tenant_quota.h [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    TenantQuotaResult ReleasePartial(std::string tenant_id, uint64_t bytes);
+
+   private:
+    TenantQuotaState& GetOrCreateState(const std::string& tenant_id);
+    TenantQuotaSnapshot MakeSnapshot(const std::string& tenant_id,
+                                     const TenantQuotaState& state) const;
+    void RefreshOverQuota(TenantQuotaState* state) const;
+
+    uint64_t default_requested_quota_bytes_ = 0;
+    std::map<std::string, TenantQuotaState> tenants_;
```
[the remark this answers is not in the document]
Comment (Lin-z-w) [reply mooncake-store/include/tenant_quota.h:70]: Resolving this together with the map/unordered_map discussion above. Since `TenantQuotaTable` intentionally keeps `std::map` in PR1, no include change is needed.

File: mooncake-store/src/tenant_quota.cpp [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+std::vector<TenantQuotaSnapshot> TenantQuotaTable::ListTenantSnapshots() const {
+    std::vector<TenantQuotaSnapshot> snapshots;
+    for (const auto& [tenant_id, state] : tenants_) {
+        if (IsLazyEmptyTenant(state)) {
+            continue;
+        }
+        snapshots.push_back(MakeSnapshot(tenant_id, state));
+    }
+    return snapshots;
+}
```
[the remark this answers is not in the document]
Comment (Lin-z-w) [reply mooncake-store/src/tenant_quota.cpp:180]: Thanks. Since this PR keeps `std::map`, `ListTenantSnapshots()` already returns snapshots in `tenant_id` order. If we later switch the backing container to `unordered_map`, we should add explicit sorting at that point.

File: mooncake-store/src/tenant_quota.cpp [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+        return tl::make_unexpected(TenantQuotaError::kQuotaExceeded);
+    }
+    auto& state = it->second;
+
+    if (state.used_bytes > state.effective_quota_bytes ||
+        state.reserved_bytes > state.effective_quota_bytes - state.used_bytes ||
+        bytes > state.effective_quota_bytes - state.used_bytes -
+                    state.reserved_bytes) {
+        return tl::make_unexpected(TenantQuotaError::kQuotaExceeded);
+    }
```
[the remark this answers is not in the document]
Comment (Lin-z-w) [reply mooncake-store/src/tenant_quota.cpp]: Good point. Fixed.

File: mooncake-store/src/tenant_quota.cpp [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+TenantQuotaResult TenantQuotaTable::Commit(std::string tenant_id,
+                                           uint64_t bytes) {
+    auto normalized_tenant_id = NormalizeTenantId(std::move(tenant_id));
+    auto& state = GetOrCreateState(normalized_tenant_id);
+    if (bytes == 0) {
+        return {};
+    }
+
+    if (state.reserved_bytes < bytes) {
```
Comment (yokinoshitayoki) [inline mooncake-store/src/tenant_quota.cpp:215]: I think the accounting-mismatch paths can leave the quota state internally inconsistent. For example, if a tenant has multiple committed objects and `Release()` is called with `bytes > used_bytes`, this branch clears `used_bytes` to 0 but only decrements `committed_count` by one. `ReleasePartial()` has a similar issue because the underflow path clears `used_bytes` without touching `committed_count`, and `Commit()` with `bytes > reserved_bytes` moves the partial reservation into `used_bytes` without incrementing `committed_count`.

Could we either make accounting-mismatch paths non-mutating, or clamp all related counters together so snapshots remain self-consistent?

Comment (Lin-z-w) [reply mooncake-store/src/tenant_quota.cpp:215]: Good catch, thanks. I agree that the mismatch paths should not try to partially repair accounting state, since these cases represent caller contract violations and we cannot reliably infer the intended object/count relationship.

I updated the accounting mismatch paths to be non-mutating for `Commit`, `Abort`, `Release`, and `ReleasePartial`: they now only log and return `kAccountingMismatch`, leaving `reserved_bytes`, `used_bytes`, and `committed_count` unchanged. I also updated the unit tests to compare before/after snapshots and verify the state remains unchanged on these mismatch paths.

**Review by yokinoshitayoki — APPROVED** [review]
LGTM.
